wordpress-mobile/WordPress-iOS
# Sync authors with the Blog Service

This is one PR in a series towards implementing #9015.

This PR is branched from #16269, so it should be merged first.

**Why this change is needed:** Once we provide the ability for the user to select a new author for a Post or Page, we don't want removed authors to stay in that list.

**What changed:**

1. The `addToAuthors` and `removeFromAuthors` methods that took `NSSet` were updated for a parameter of `Set<BlogAuthor>`. This should now match their single value counterpart.
2. If our local set of authors includes a userID that doesn't exist in our latest `RemoteUser` array, we remove that author from the local store because they've been removed from the blog.

### To test:

The only area I've identified as being impacted is the History view for Posts:

1. Make changes to a WordPress.com Post as multiple users (e.g. User A and User B have both made updates to the same Post).
2. With the post presented in the app, tap the horizontal ellipses (More Options) icon at the top right.
3. Select "History".
4. Observe the user names and avatars next to each revision.
5. Remove an author from the blog.
6. Restart the app.
7. View the history for the same post.
8. Observe that the removed author no longer has a display name nor avatar icon.

## Regression Notes
1. Potential unintended areas of impact

- Any areas that expect a deleted author's information to be available. 
- Since the `authors` property is NSManaged, I would appreciate confirmation that using `Set<BlogAuthor>` over the original `NSSet` parameter won't cause any negative effects.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Used the test sequence above.

3. What automated tests I added (or what prevented me from doing so)
Added `BlogServiceAuthorTests`.

PR submission checklist:
